### PR TITLE
adtool: unbreak it

### PR DESCRIPTION
**Pull request by peterhoeg**
## Description of changes

unbreak it

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/admin/adtool/default.nix [hunk 1]
```diff
@@ -1,46 +1,67 @@
-{ lib, stdenv, fetchurl, openldap }:
+{ lib, stdenv, fetchFromGitHub, fetchurl, db, openldap }:
 
-stdenv.mkDerivation rec {
+let
+  inherit (lib) getDev getLib;
+
+  # adtools needs openldap 2.4.x
+  openldap' = openldap.overrideAttrs (old: rec {
+    version = "2.4.59";
```
Comment (Ma27) [inline pkgs/tools/admin/adtool/default.nix:8]: I'm not sure if that's a good idea: not only that you can't modify this openldap' any further from outside. It's also a two year old version that doesn't receive any security updates anymore.

**Comment by peterhoeg** [comment]
As `adtool' is a client only, in order for someone to exploit any issues in this version of openldap, you would need to control a server that can send malicious content to `adtool', so the risk seems extremely limited IMHO.

That being said, there might be CVEs applicable to this version- I admittedly haven’t checked. So we could obviously tag this accordingly if there are any. *b*, even if there is, I don’t plan on reviewing the entire adtool code base to see if it actually uses the vulnerable parts or not.

I fully understand where you are coming from, I just thing the risk is absolutely managable given this version being restricted to `adtool' only *and* adtool being a client only.

Thoughts?

File: pkgs/tools/admin/adtool/default.nix (hunk 1 above)
Comment (symphorien) [reply pkgs/tools/admin/adtool/default.nix:8]: well I suppose if we go this route it should get meta.knownVulnerabilities

Comment (Ma27) [reply pkgs/tools/admin/adtool/default.nix:8]: > well I suppose if we go this route it should get meta.knownVulnerabilities

That's the least thing we should do, but I'm not convinced that we should revive dead software on unstable unless there's a very good reason.
Anyways, forwarded it to security discussion to get some more opinions.

**Comment by peterhoeg** [comment]
> but I’m not convinced that we should revive dead software on unstable unless there’s a very good reason.

Software that still works and is used by someone isn’t really dead IMHO despite the upstream development speed being slow (or not moving at all).

And again, there’s a huge difference between exposing software with known vulnerabilities to a wider audience when run as a network reachable service and using a client that uses a possibly vulnerable component.

> Anyways, forwarded it to security discussion to get some more opinions.

Do you have a link? It’s absolutely a “bigger” and more principled discussion than fixing this specific tool.

**Comment by Ma27** [comment]
> Software that still works and is used by someone isn’t really dead IMHO despite the upstream development speed being slow (or not moving at all).

I hope no Windows XP user follows this thread.

> And again, there’s a huge difference between exposing software with known vulnerabilities to a wider audience when run as a network reachable service and using a client that uses a possibly vulnerable component.

For a distribution there's an expectation that we ship a secure package-set. This actively breaks this assumption and this would also set a problematic precedence for similar changes in the future. It's unfortunate enough that we had to keep e.g. python2 after its EOL for so long.

So to me this sounds like something that belongs to an unofficial overlay.

**Comment by peterhoeg** [comment]
I hope no Windows XP user follows this thread.




I’m not entirely sure what point you are trying to make here. WinXP still has its uses and as long as necessary precautions are taken, that may very well be the best solution. Putting it on an end-user device used to access the wider internet? Not so great. Controlling industrial machinery in an air-gapped environment? Might be your only option.




For a distribution there’s an expectation that we ship a secure package-set.




“Secure” is not a binary choice between “secure” and “not secure”. Security is on a continuum.




This actively breaks this assumption and this would also set a problematic precedence for similar changes in the future.




Shipping adtool means “here is a piece of software that might expose you to having your device compromised if you are using it to talk to a domain controller controlled by a malicious entity”. If someone can plop a new DC on your network, you have waaaaaay bigger problems than an old version of openldap. Realistically, what are the odds that you would be impacted by this? It’s not 0, but it’s pretty damn low. You don’t use it to connect to random domain controllers - it’s mainly used for internal networks.




It’s unfortunate enough that we had to keep e.g. python2 after its EOL for so long.




That is a completely different beast. A language interpreter and an entire ecosystem with thousands of plugins used to write all kinds of software doing all kinds of things exposed to all kinds of hostilities vs a tool used for talking to internal, trusted servers. Apples to oranges.




So to me this sounds like something that belongs to an unofficial overlay.




That might very well be - in any case, it’s a discussion worth having as it is of a principle nature. Did you raise it elsewhere? If not, I’ll go ahead.

**Comment by Ma27** [comment]
> Controlling industrial machinery in an air-gapped environment? Might be your only option

Sure, but such a use-case doesn't qualify for inclusion into a distribution. With that reasoning you could keep (almost) any package in nixpkgs, including python2.

If you need to use that tool, that's perfectly fine, I don't mind. My point is that I'm not convinced that this should be part of nixpkgs then.

> Did you raise it elsewhere? If not, I’ll go ahead.

I asked in the NixOS security discussion Matrix channel, but didn't get an answer there, so go ahead.

Part of my reservation is that we'd probably set a precedence with this patch that might bite us in the future. So yeah, discussing this in general and not about this package in particular is a good way forward.

**Pull request closed without merging**